examples/Inference.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, TypeApplications, TypeOperators, UndecidableInstances #-}
+module Inference
```

Comment:
I brought this over from the unit tests and expanded on it significantly to document the idiom.